### PR TITLE
[WIP] Initial support for sparse labels on confusion-matrix metrics

### PR DESCRIPTION
**Pull request by dwyatte**
This PR addresses https://github.com/tensorflow/tensorflow/issues/37104 by supporting sparse labels in confusion-matrix metrics (`Precision`, `Recall`, `SensitivityAtSpecificity`, `SpecificityAtSensitivity`, etc).

Currently, the implementation simply infers whether the `y_true` argument to `update_confusion_matrix_variables` is sparse (and only when passing `class_id`). However, I think I'd personally prefer something more explicit since this enables users to call metrics like

```
m = tf.keras.metrics.Precision(class_id=3)
m.update([3, 0, 0], [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0]])
```

where `y_true` (first arg) and `y_pred` (second arg) have different dimensions (but the same length) which is something other metrics don't allow, to my knowledge.

Two ideas:
1. Pass an arg to the metric constructor to indicate that `y_true` is sparse like `tf.keras.metrics.Precision(class_id=3, sparse_labels=True)`
2. Create Sparse variants of metrics like `SparsePrecision`, `SparseRecall`, etc. This is the current pattern used for accuracy (`SparseCategoricalAccuracy`, `SparseCategoricalCrossentropy`, etc.)

I think I slightly prefer 1. as it is a little more lightweight compared to proliferating classes/test cases. However perhaps others don't think we need either of these options and can just simply infer whether the `y_true` argument is sparse as in the current iplementation.

### CODE REVIEW

**Comment by gbaned** [comment]
@dwyatte This PR is in draft, any update on this? Please. Thanks!

**Comment by dwyatte** [comment]
@gbaned I opened this as a draft as I was hoping to get some guidance from core contributors/maintainers on whether implicitly inferring sparse labels based on the size of the `y_true`/`y_pred` args to a metric is sufficient or whether one of the alternate designs mentioned in https://github.com/tensorflow/tensorflow/pull/48122#issue-602111321 is preferred. 

What do you think? I can convert to non-draft if easier to iterate that way. 

Certainly I think this is worth addressing as it frequently comes up here (https://github.com/tensorflow/tensorflow/issues/37104, https://github.com/tensorflow/tensorflow/issues/42383) and on StackOverflow ([1](https://stackoverflow.com/questions/61742556/valueerror-shapes-none-1-and-none-2-are-incompatible, [2](https://datascience.stackexchange.com/questions/73827/valueerror-shapes-are-incompatible-when-fitting-using-imagedatagenerator)).

**Comment by qlzh727** [comment]
Sorry for the late reply. Looking at the issue, I would prefer to have separate metric class like SparsePrecision/SparseRecall to handle the sparse label case. The semantic between the SparsePrecision and Precision is actually a bit different, eg the Precision has a param threshold to control when a prediction is considered true/false. In the sparse case, the index of argmax will become the result and compare against the label.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
